src/Components/test/testassets/BasicTestApp/VirtualizationComponent.razor

```diff
@@ -0,0 +1,74 @@
+﻿<p>
+    Item size:<br />
+    <input id="item-size-input" type="number" @bind-value="itemSize" />
+</p>
+
+<p>
+    Synchronous:<br />
+    <div id="sync-container" style="background-color: #eee; height: 500px; overflow-y: auto">
+        <Virtualize Items="@fixedItems" ItemSize="itemSize">
+            <div id="sync-item" style="height: @(itemSize)px; background-color: rgb(@((context % 2) * 255), @((1-(context % 2)) * 255), 255);">Item @context</div>
+        </Virtualize>
+    </div>
+</p>
+
+<p>
+    Asynchronous:<br />
+    <button id="finish-loading-button" @onclick="FinishLoadingAsync">Finish loading</button><br />
+    Cancellation count: <span id="cancellation-count">@asyncCancellationCount</span><br />
+    <div id="async-container" style="background-color: #eee; height: 500px; overflow-y: auto">
+        <Virtualize ItemsProvider="GetItemsAsync" ItemSize="itemSize">
+            <Item>
+                <div id="async-item" style="height: @(itemSize)px; background-color: rgb(@((context % 2) * 255), @((1-(context % 2)) * 255), 255);">Item @context</div>
+            </Item>
+            <Placeholder>
+                <div id="async-placeholder" style="height: @(itemSize)px; background-color: orange;">Loading item @context.Index...</div>
+            </Placeholder>
+        </Virtualize>
+    </div>
+</p>
+
+<p id="viewport-as-root">
+    Viewport as root:<br />
+    <Virtualize Items="@fixedItems" ItemSize="itemSize">
+        <div style="height: @(itemSize)px; background-color: rgb(@((context % 2) * 255), @((1-(context % 2)) * 255), 255);">Item @context</div>
+    </Virtualize>
+</p>
+
+@code {
+    float itemSize = 100;
+    ICollection<int> fixedItems = Enumerable.Range(0, 1000).ToList();
+
+    int asyncTotalItemCount = 200;
+    int asyncCancellationCount = 0;
+    TaskCompletionSource asyncTcs = new TaskCompletionSource();
+
+    HashSet<int> cachedItems = new HashSet<int>();
+
+    async ValueTask<ItemsProviderResult<int>> GetItemsAsync(ItemsProviderRequest request)
+    {
+        var registration = request.CancellationToken.Register(() => CancelLoadingAsync(request.CancellationToken));
```

Comment:
If request.CancellationToken is already cancelled then CancelLoadingAsync will be invoked inline and replace asyncTcs with a new instance before you await it. Capture asyncTcs locally before calling Register.